designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,336 @@
+
+Debug Configurations are stored in `.vscode/launch.json` relative to the VS Code workspace.
+
+More information about VS Code Debugging can be found [in the VS Code Documentation](https://code.visualstudio.com/docs/editor/debugging).
+
+#### <a id="terms-sam-template"></a> SAM Template
+
+A SAM Template defines a Serverless Application's resources, and supporting code. The SAM CLI provides tooling around this template model to build, run, package, and deploy the Application.
+
+Additional information about SAM can be found at:
+
+-   [SAM Homepage](https://aws.amazon.com/serverless/sam/)
+-   [What Is the AWS Serverless Application Model (AWS SAM)?](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam.html) (includes a Getting Started guide to the SAM CLI)
+-   [SAM CLI GitHub Repo](https://github.com/awslabs/aws-sam-cli)
+
+### Differences between this design, and v1.0.0 (through present versions) of AWS Toolkit
+
+The debug capabilities initially released in the Toolkit were not well rounded. CodeLenses provided the only means of local debugging, and did not have a well defined design around what was being debugged. Because the Toolkit lacked a way to directly invoke/debug SAM Template resources, these CodeLenses tried to compensate, leading to scenarios with undefined behaviors. Many of these issues are referenced from https://github.com/aws/aws-toolkit-vscode/issues/758
+
+Here is an outline of the differences between this design and the current version of the AWS Toolkit:
+
+-   New functionality
+    -   `aws-sam` Debug Configurations provide a way to launch debug sessions that articulate what is being debugged
+    -   Added CodeLenses to help users to produce Debug Configurations
+-   Removed functionality
+    -   Using CodeLenses as a mechanism to launch debug sessions will be removed. Users familiar with the outgoing functionality are provided with an assisted transition path towards using Debug Configurations.
+        -   the toolkit detects if users were using the previous CodeLens functionality based on the presence of a CodeLens configuration file (`{workspace}/.aws/templates.json`). For these users, lambda handlers that used to show Run/Debug CodeLenses will be decorated to mention the functionality shift, and will offer to convert the function's old CodeLens configuration over to a new debug configuration.
+
+### <a id="sample-debug-configurations"></a> Sample Debug Configurations
+
+Configuration structures are modelled so that they group related parameters, and allows the toolkit to reuse structure types where possible.
+
+Here is an example Debug Configuration to debug a SAM Template resource called "HelloWorldResource".
+The required fields are: type, request, invokeTarget
```

Comment:
`type` and `request` are always required fields, with `request` dictating required parameters below (specifically: the `request` type is dictated by the keys of the objects directly nested under `configurationAttributes` here: https://code.visualstudio.com/api/references/contribution-points#debugger-example )

Because of this, we cannot have the same `request` type with different required parameters (see: https://github.com/aws/aws-toolkit-vscode/pull/933/files#diff-d1b7a9dd4e20bef11ef6114f80e8d1dcR229)

Could you please provide different `request` names?

Reply:
We should be able to accommodate this. If not, the doc can be updated during implementation, prior to being available.

Reply:
This isn't validatable via package.json but we're opting to drive the required fields through the DebugConfigurationProvider which will validate instead.